theano/sandbox/mkl/mkl_type.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import absolute_import, print_function, division
+import os
+import numpy
+import theano
+from theano import Type, Variable
+from theano import tensor
+from six import StringIO
+
+try:
+    import mkl_ndarray
+    import mkl_ndarray.mkl_ndarray as mkl
+
+    try:
+        mkl_ndarray.__file__
+    except AttributeError:
+        from theano.gof.cmodule import get_lib_extension
+        mkl_ndarray.__file__ = os.path.join(mkl_ndarray.__path__._path[0], 'mkl_ndarray.' + get_lib_extension())
+except ImportError:
+    mkl = None
+
+
+class _operators(tensor.basic._tensor_py_operators):
+    dtype = property(lambda s: s.type.dtype)
+    broadcastable = property(lambda s: s.type.broadcastable)
+    ndim = property(lambda s: s.type.ndim)
+
+
+class MKLNdarrayVariable(_operators, Variable):
+    pass
+
+
+class MKLNdarrayType(Type):
```

Comment:
Let me introduce some basic background about Intel MKL DNN primitives firstly.
 
Intel MKL DNN primitives work with special opaque data types such as operation primitives (such as convolution forward/backward primitive, etc.,) and layouts of processed data (specifies relative location of elements of processed arrays in memory). Each DNN operation requires input and output arrays have certain data layouts.
 
More detail information can be found [here](https://software.intel.com/en-us/node/684759)
 
 
The reasons why we added such a new C object are: 
 
1.	Numpy ndarray and GPU ndarray doesn’t support storing extra variable such as layout structure.
The memory view described by Numpy ndarray and GPU ndarray is via dims and strides, while it’s described by an opaque data types (layouts) in MKL.
Besides, layouts needs to be passed from one Op to the next one for the computation. We need an C object (TensorType) which allows us to pass the layout.

2.	The data buffer of Numpy ndarray and GPU ndarray can’t be reused by MKL due to different size.
The size of data buffer of Numpy ndarray is exact same as described by dims and strides. While the size of data buffer of mkl op is described by its layout, and it might have different size compared with numpy (might include padding)
 
3.	Same as item 1.
For some Ops, backward pass needs the intermediate result which calculated in forward pass. So we need another data buffer named "workspace" in the variable.
 
Actually, in the beginning, we're thinking about leveraging Numpy ndarray or Cuda/GPU ndarray for this purpose, but none of them allows us to do so. So that's why we create a simplified but sufficient structure for mkl Ops. 

We're glad to see If there's any better suggestions.